rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,164 @@
+collapse an (encoder) transform + sink into a sink with encoder config, in a way
+that is transparent to the user. These transforms merely convert between
+encoding formats, e.g. from `bytes` to `json` and vice versa.
+
+## Scope
+
+The scope of this RFC concern at which level framing and decoding operates, how
+these framers and codecs can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for framing or a
+codec. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and codecs, such that each source and sink may implement them in their
+own way. This can introduce inconsistencies and additional maintenance burden.
+Components have their own defaults, making behavior unpredictable and surprising
+users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+Since decoding and encoding should apply universally to all sources and sinks,
+it should be configurable by a shared field. This is accomplished by the
+`SourceOuter` and `SinkOuter` wrappers:
+
+```rust
+// Newly introduced wrapper, analogously to `SinkOuter`.
+struct SourceOuter {
+    framing: Framer,
+    decoding: Decoder,
+}
+```
+
+```rust
+struct SinkOuter {
+    // … existing fields
+    encoding: Encoder,
+}
+```
+
+Conceptually, decoding is applied _after_ a source has processed an event and
+encoding is applied _before_ a sink has processed an event. These can be
+implemented as `FunctionTransform`s in our pipeline, in the same way transforms
+are implemented.
+
+Framing takes place before an event is created since the message boundary
+required to form an event is not known yet. Therefore the source needs to be
+aware of framing and call the framer to determine message boundaries before
+creating an event.
+
+To make the source aware of framing, the `Framer` needs to be passed to the
+`SourceContext`, so that the source implementation can call it:
+
+```rust
+struct SourceContext {
+    // … existing fields
+    framing: Framer,
+}
+```
+
+In the context of Vector, decoders and encoders are implemented as `Transform`.
+That is, they either implement the `TaskTransform` or `FunctionTransform` trait.
+
+Implementing framers as `Transform` would be possible - however, we want to
+restrict them to `FunctionTransform` at the moment. The reason here is that
+`TaskTransform`s can take any amount of input bytes before they output frames.
+If we would allow that, it wouldn't be clear how we merge metadata from multiple
+packages that compose a frame, e.g. the sender in UDP datagrams or partition key
+and offset in kafka messages. Using `FunctionTransform`, we know that a 1:n
+relationship between byte chunks and frames exist, and we can simply duplicate
+the metadata for each frame.
```

Comment:
Can `FunctionTransform`s do that? I thought they were intended to be stateless.

Reply:
Yeah I had a big misconception there. For Kafka the "1:n" model would be a workable limitation, but this wouldn't work at all for stream-based transports.

E.g. when we have a line-codec for TCP, we can't just assume that the next read will split exactly at the message boundary. Therefore it would definitely need to keep internal state.

Reply:
Most do tend to be stateless, but the actual distinction between function and task is whether a transform needs to be able delay or schedule output independent of its input. Function transforms can mutate internal state (i.e. `&mut self`), but they have no means of doing any kind of async operation internally (e.g. timing out an aggregation).

Reply:
Aha, I see, I wasn't aware that that is the distinction that is being made between them 👍

I think we have a couple of `TaskTransform`s that don't match that: `dedupe`, `tag_cardinatily_limit`, and `aws_ec2_metadata`. I wonder if they could be `FunctionTransform`s instead.

In the `dedupe`, and `tag_cardinality_limit` cases, it seems like we'd really only want one of them running; that is, we wouldn't want to clone them and push them up to the source to take advantage of the source concurrency. I think like we may still have a need to differentiate transforms that need to have a global view from those that don't. We seem to currently use `Function` vs `Task` for that distinction, but seemingly we may need a `StatelessFunctionTransform`, perhaps, that takes `&self`, whereby we'd know it is safe to clone and apply concurrently.

Reply:
Yeah, those were left as tasks when we originally separated out function transforms for that exact reason, since the original motivation of differentiating between the two was for running transforms inline in sources. Ideally we'd have a clearer distinction between the different concerns (e.g. some kind of "I can run in parallel" flag to control inlining).